File: analyzer-direct/src/test/scala/io/github/tassiLuca/analyzer/core/AnalyzerTest.scala
package io.github.tassiLuca.analyzer.core

import gears.async.default.given
import gears.async.Async
import io.github.tassiLuca.analyzer.commons.lib.{Contribution, Release, Repository, RepositoryReport}
import io.github.tassiLuca.analyzer.lib.{Analyzer, GitHubService}
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class AnalyzerTest extends AnyFlatSpec with Matchers {

  val mockGitHubService: GitHubService = new GitHubService:
    override def repositoriesOf(organizationName: String)(using Async): Either[String, Seq[Repository]] =
      Right(Seq(Repository(0, "dse/test-1", 100, 10), Repository(1, "dse/test-2", 123, 198)))

    override def contributorsOf(
        organizationName: String,
        repositoryName: String,
    )(using Async): Either[String, Seq[Contribution]] = repositoryName match
      case "test-1" => Right(Seq(Contribution("mrossi", 56), Contribution("lpluto", 11)))
      case "test-2" => Right(Seq(Contribution("mrossi", 11), Contribution("averdi", 98)))

    override def lastReleaseOf(organizationName: String, repositoryName: String)(using Async): Either[String, Release] =
      repositoryName match
        case "test-1" => Right(Release("v0.1", "2024-02-21"))
        case _ => Left("404, not found")

  val analyzer: Analyzer = Analyzer.of(mockGitHubService)

  val expectedResults = Seq(
    RepositoryReport(
      "test-1",
      10,
      100,
      Seq(Contribution("mrossi", 56), Contribution("lpluto", 11)),
      Some(Release("v0.1", "2024-02-21")),
    ),
    RepositoryReport(
      "test-2",
      198,
      123,
      Seq(Contribution("mrossi", 11), Contribution("averdi", 98)),
      None,
    ),
  )

  "simple test" should "work" in {
    var results: Set[RepositoryReport] = Set()
    Async.blocking:
      val allResults = analyzer.analyze("dse") { report =>
        results += report
      }
      println(allResults)
      allResults.isRight shouldBe true
      allResults.map(_ should contain theSameElementsAs expectedResults)
      results.toSeq should contain theSameElementsAs expectedResults
  }
}

File: analyzer-direct-kt/src/test/kotlin/io/github/tassiLuca/analyzer/lib/AnalyzerTest.kt
package io.github.tassiLuca.analyzer.lib

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldContainAll
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.runBlocking
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`

abstract class AnalyzerTest : FunSpec() {

    protected val dummiesData: Map<Repository, Pair<Set<Contribution>, Release?>> = mapOf(
        Repository(0, "dse/test-1", 100, 10) to
            Pair(setOf(Contribution("mrossi", 56)), Release("v0.1", "2024-02-21")),
        Repository(1, "dse/test-2", 123, 198) to
            Pair(setOf(Contribution("mrossi", 11), Contribution("averdi", 98)), null),
    )

    init {
        test("Channel based analyzer should return the correct results if no errors occur") {
            var incrementalResults = emptySet<RepositoryReport>()
            runBlocking {
                val service = successfulService()
                val allResults = service.analyze("dse") {
                    incrementalResults = incrementalResults + it
                }
                allResults.isSuccess shouldBe true
                incrementalResults.isEmpty() shouldBe false
                allResults.getOrThrow() shouldContainAll expectedResults()
            }
        }

        test("Channel based analyzer should return a failure if given in input a non-existing organization") {
            var incrementalResults = emptySet<RepositoryReport>()
            runBlocking {
                val service = failingService()
                val allResults = service.analyze("dse") {
                    incrementalResults = incrementalResults + it
                }
                allResults.isFailure shouldBe true
                incrementalResults.shouldBeEmpty()
            }
        }
    }

    private fun expectedResults() = dummiesData.map {
        RepositoryReport(it.key.name, it.key.issues, it.key.stars, it.value.first.toList(), it.value.second)
    }.toSet()

    abstract fun analyzer(provider: GitHubRepositoryProvider): Analyzer

    private suspend fun successfulService(): Analyzer {
        val gitHubProvider = mock<GitHubRepositoryProvider>()
        registerSuccessfulRepositoriesResult(gitHubProvider)
        dummiesData.forEach { (repo, data) ->
            `when`(gitHubProvider.lastReleaseOf(repo.organization, repo.name))
                .thenReturn(Result.success(data.second))
        }
        return analyzer(gitHubProvider)
    }

    abstract suspend fun registerSuccessfulRepositoriesResult(provider: GitHubRepositoryProvider)

    private suspend fun failingService(): Analyzer {
        val gitHubProvider = mock<GitHubRepositoryProvider>()
        registerFailingRepositoriesResult(gitHubProvider)
        return analyzer(gitHubProvider)
    }

    abstract suspend fun registerFailingRepositoriesResult(provider: GitHubRepositoryProvider)
}

class GitHubAnalyzerByChannelsTest : AnalyzerTest() {
    override fun analyzer(provider: GitHubRepositoryProvider): Analyzer = Analyzer.ofGitHubByChannels(provider)

    override suspend fun registerSuccessfulRepositoriesResult(provider: GitHubRepositoryProvider) {
        `when`(provider.repositoriesOf("dse"))
            .thenReturn(Result.success(dummiesData.keys.toList()))
        dummiesData.forEach { (repo, data) ->
            `when`(provider.contributorsOf(repo.organization, repo.name))
                .thenReturn(Result.success(data.first.toList()))
        }
    }
    override suspend fun registerFailingRepositoriesResult(provider: GitHubRepositoryProvider) {
        `when`(provider.repositoriesOf("dse"))
            .thenReturn(Result.failure(RuntimeException("404, not found")))
    }
}

class GitHubAnalyzerByFlowsTest : AnalyzerTest() {
    override fun analyzer(provider: GitHubRepositoryProvider): Analyzer = Analyzer.ofGithubByFlows(provider)

    override suspend fun registerSuccessfulRepositoriesResult(provider: GitHubRepositoryProvider) {
        `when`(provider.flowingRepositoriesOf("dse"))
            .thenReturn(dummiesData.keys.asFlow().map { println(it); listOf(it) })
        dummiesData.forEach { (repo, data) ->
            `when`(provider.flowingContributorsOf(repo.organization, repo.name))
                .thenReturn(data.first.asFlow().map { listOf(it) })
        }
    }
    override suspend fun registerFailingRepositoriesResult(provider: GitHubRepositoryProvider) {
        `when`(provider.flowingRepositoriesOf("dse"))
            .thenReturn(flow { throw RuntimeException("404, not found") })
    }
}
